legacy/firmware/protect.c

```diff
@@ -369,12 +370,25 @@ bool protectPassphrase(void) {
   bool result;
   for (;;) {
     usbPoll();
-    // TODO: correctly process PassphraseAck with state field set (mismatch =>
-    // Failure)
     if (msg_tiny_id == MessageType_MessageType_PassphraseAck) {
       msg_tiny_id = 0xFFFF;
       PassphraseAck *ppa = (PassphraseAck *)msg_tiny;
-      session_cachePassphrase(ppa->has_passphrase ? ppa->passphrase : "");
+      if (ppa->has_on_device && ppa->on_device == true) {
+        fsm_sendFailure(
+            FailureType_Failure_DataError,
+            _("This firmware is incapable of passphrase entry on the device."));
+        result = false;
+        break;
+      }
+      if (!ppa->has_passphrase) {
+        fsm_sendFailure(FailureType_Failure_DataError,
+                        _("No passphrase provided. Use empty string to set an "
+                          "empty passphrase."));
+        result = false;
+        break;
+      }
+      // TODO: ask - why ppa->passphrase.size is not working? because of tiny?
+      strlcpy(passphrase, ppa->passphrase, sizeof(ppa->passphrase));
```

Comment:
cc @prusnak

Reply:
No. It doesn't work because ppa->passphrase is of type `string`, not `bytes`. Strings in nanopb are zero-terminated `char *`, Bytes in nanopb are `struct { pb_size_t size; pb_byte_t bytes[n]; }`